From 249383abbff87f11b2b1eac1fe2f5ee72910faf3 Mon Sep 17 00:00:00 2001
From: Don Zickus <dzickus@redhat.com>
Date: Tue, 16 Sep 2014 17:10:01 -0400
Subject: [usb] xhci: Fix sleeping with IRQs disabled in xhci_stop_device()

Message-id: <1410887463-20674-530-git-send-email-dzickus@redhat.com>
Patchwork-id: 95681
O-Subject: [RHEL7 PATCH 529/591] xhci: Fix sleeping with IRQs disabled in xhci_stop_device()
Bugzilla: 1110939
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1110939

commit be3de32107091c266b00a48265fe7e06233af4f0
Author: Mathias Nyman <mathias.nyman@linux.intel.com>
Date:   Tue Jun 10 11:27:41 2014 +0300

    xhci: Fix sleeping with IRQs disabled in xhci_stop_device()

    xhci_stop_device() allocates and issues stop commands for each active endpoint.
    This is done with spinlock held and interrupt disabled so we can't sleep during
    memory allocation. Use GFP_NOWAIT instead

    Regression from commit ddba5cd0aeff5bbed92ebdf4b1223300b0541e78
    "xhci: Use command structures when queuing commands on the command ring"
    for 3.16-rc1

    Fixes: ddba5cd0aeff ("xhci: Use command structures when queuing commands")
    Reported-by: Dan Williams <dan.j.williams@intel.com>
    Reported-by: Johannes Berg <johannes@sipsolutions.net>
    Signed-off-by: Mathias Nyman <mathias.nyman@linux.intel.com>
    Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/usb/host/xhci-hub.c b/drivers/usb/host/xhci-hub.c
index 6231ce6..2b998c6 100644
--- a/drivers/usb/host/xhci-hub.c
+++ b/drivers/usb/host/xhci-hub.c
@@ -287,7 +287,7 @@ static int xhci_stop_device(struct xhci_hcd *xhci, int slot_id, int suspend)
   if (virt_dev->eps[i].ring && virt_dev->eps[i].ring->dequeue) {
    struct xhci_command *command;
    command = xhci_alloc_command(xhci, false, false,
-           GFP_NOIO);
+           GFP_NOWAIT);
    if (!command) {
     spin_unlock_irqrestore(&xhci->lock, flags);
     xhci_free_command(xhci, cmd);
-- 
1.7.1